### PR TITLE
feat: event aggregator

### PR DESCRIPTION
**Pull request by carlos-r-l-rodrigues**
What:
- Event Aggregator Util
- Preparation for normalizing event in a new format (backward compatible with the current format)
- GQL Schema to joiner config and some Entities configured
- Link modules emmiting events

### CODE REVIEW

File: packages/product/src/services/product-module-service.ts [hunk 1]
```diff
@@ -1239,6 +1297,11 @@ export default class ProductModuleService<
         productData.options,
         sharedContext
       )
+
+      console.log(
+        "++++++productData.options++++++++++++++++++++++++++++++",
+        productData.options
+      )
```
Comment (olivermrbl) [inline packages/product/src/services/product-module-service.ts]:
```suggestion
```

File: packages/medusa/src/services/product-variant-inventory.ts [hunk 2]
```diff
@@ -404,7 +432,24 @@ class ProductVariantInventoryService extends TransactionBaseService {
...
+    const createdVariantInventoryItems = await variantInventoryRepo.save(
+      toCreate
+    )
+
+    await this.eventBusService_.emit(
+      buildEventMessages({
+        // TODO: align the event entity format later
+        eventName: "LinkProductVariantInventoryItem.attached",
+        data: createdVariantInventoryItems.map((v) => ({ id: v.id })),
+        metadata: {
```
Comment (olivermrbl) [inline packages/medusa/src/services/product-variant-inventory.ts]: question: For the common case, do you think we could autogenerate the metadata?

Here it looks like 
- `object === eventName.split(".")[0]`
- `service === snakeCase(this.constructor.name)`
- `action === eventName.split(".")[1]`


I don't know enough about the metadata, but maybe there is a pattern so we don't have to include this every time.

File: packages/link-modules/src/services/link-module-service.ts [hunk 3]
```diff
@@ -297,14 +362,29 @@ export default class LinkModuleService<TLink> implements ILinkModule {
       )
     }
 
+    await this.eventBusModuleService_?.emit<Record<string, unknown>>(
+      (restoredEntities as { id: string }[]).map(({ id }) => ({
+        eventName: this.entityName_ + ".attached",
+        body: {
+          metadata: {
```
Comment (olivermrbl) [inline packages/link-modules/src/services/link-module-service.ts:370]: See other comment. I was thinking something like this for emitting events in other modules.

File: packages/link-modules/src/services/link-module-service.ts [hunk 4]
```diff
@@ -256,14 +303,29 @@ export default class LinkModuleService<TLink> implements ILinkModule {
       )
     }
 
+    await this.eventBusModuleService_?.emit<Record<string, unknown>>(
+      (deletedEntities as { id: string }[]).map(({ id }) => ({
+        eventName: this.entityName_ + ".dettached",
+        body: {
+          metadata: {
+            service: this.serviceName_,
+            action: "dettached",
```
Comment (olivermrbl) [inline packages/link-modules/src/services/link-module-service.ts]: todo: Typo, perhaps these actions should be "constified"
```suggestion
        eventName: this.entityName_ + ".detached",
        body: {
          metadata: {
            service: this.serviceName_,
            action: "detached",
```

File: packages/link-modules/src/services/link-module-service.ts [hunk 5]
```diff
@@ -224,6 +252,22 @@ export default class LinkModuleService<TLink> implements ILinkModule {
...
     await this.linkService_.delete(data, sharedContext)
+
+    const allData = Array.isArray(data) ? data : [data]
+    await this.eventBusModuleService_?.emit<Record<string, unknown>>(
+      allData.map(({ id }) => ({
+        eventName: this.entityName_ + ".dettached",
+        body: {
+          metadata: {
+            service: this.serviceName_,
+            action: "dettached",
```
Comment (olivermrbl) [inline packages/link-modules/src/services/link-module-service.ts]:
```suggestion
        eventName: this.entityName_ + ".detached",
        body: {
          metadata: {
            service: this.serviceName_,
            action: "detached",
```

File: packages/utils/src/event-bus/common-events.ts [hunk 6]
```diff
@@ -0,0 +1,7 @@
+export enum CommonEvents {
+  CREATED = "created",
+  UPDATED = "updated",
+  DELETED = "deleted",
+  ATTACHED = "attached",
+  DETTACHED = "dettached",
```
Comment (olivermrbl) [inline packages/utils/src/event-bus/common-events.ts]:
```suggestion
  DETACHED = "detached",
```

File: packages/medusa/src/services/product-variant-inventory.ts (hunk 2 above)
Comment (carlos-r-l-rodrigues) [reply packages/medusa/src/services/product-variant-inventory.ts]: yes, absolutely. once we have a defined pattern we can auto generate these.
to me the event name should be `module.entity.action`.

This event emit has to be removed from this PR.
All event normalization and firing them following the new format will be part of another PR in the future.
Only link modules will fire the events in this PR

File: packages/link-modules/src/services/link-module-service.ts [hunk 7]
```diff
@@ -188,6 +202,21 @@ export default class LinkModuleService<TLink> implements ILinkModule {
 
     const links = await this.linkService_.create(data, sharedContext)
 
+    await this.eventBusModuleService_?.emit<Record<string, unknown>>(
+      (data as { id: unknown }[]).map(({ id }) => ({
+        eventName: this.entityName_ + ".attached",
```
Comment (olivermrbl) [inline packages/link-modules/src/services/link-module-service.ts]: nit: Use `CommonEvents`
```suggestion
        eventName: this.entityName_ + "." + CommonEvents.ATTACHED,
```

File: packages/link-modules/src/services/link-module-service.ts [hunk 8]
```diff
@@ -297,14 +363,29 @@ export default class LinkModuleService<TLink> implements ILinkModule {
       )
     }
 
+    await this.eventBusModuleService_?.emit<Record<string, unknown>>(
+      (restoredEntities as { id: string }[]).map(({ id }) => ({
+        eventName: this.entityName_ + ".attached",
```
Comment (olivermrbl) [inline packages/link-modules/src/services/link-module-service.ts]:
```suggestion
        eventName: this.entityName_ +  "." + CommonEvents.ATTACHED,
```

**Review by olivermrbl — APPROVED** [review]

File: packages/product/integration-tests/__tests__/services/product-module-service/products.spec.ts [hunk 9]
```diff
@@ -578,13 +578,29 @@ describe("ProductModuleService products", function () {
 
     afterEach(afterEach_)
 
-    it("should create a product", async () => {
+    it.only("should create a product", async () => {
```
Comment (pKorsholm) [inline packages/product/integration-tests/__tests__/services/product-module-service/products.spec.ts]:
```suggestion
    it("should create a product", async () => {
```

**Review by pKorsholm — APPROVED** [review]
lgtm!

**Review by adrien2p — APPROVED** [review]
Nice addition, we are starting to get all the things from our search pr 💪

File: packages/product/integration-tests/__tests__/services/product-module-service/products.spec.ts (hunk 9 above)
Comment (carlos-r-l-rodrigues) [reply packages/product/integration-tests/__tests__/services/product-module-service/products.spec.ts]: thanks. nice catch

**Pull request merged**